[draft] Bindings (#316)
solidjs/solid-docs-next: content/essentials/refs.mdx

> # Refs

Author: @modderme123

can you review this page

# Refs

Refs, or references, allow you to have direct access to DOM element. 

> To create a `ref` in Solid, you can get a reference to an element through assigning it to avariable: 

Reviewer: ```suggestion
There are two ways to get a reference to a DOM element in Solid:
1. Using JSX as a value, and assigning it to a variable
2. Using the Solid `ref` attribute
```

```jsx
const myDiv = <div>My Element</div>;
```

> For a more idiomatic way, Solid provides the `ref` attribute. 

Reviewer: Aside: I believe the reason this is more idiomatic is because it is often nice to partially separate your DOM code from your data code

Refs in Solid are assignments that occur at creation time before elements are attached to the document DOM.
To use this, you simply need to declare a variable and use it as a ref attribute:

```tsx

> let myDiv;

Reviewer: Probably worth pointing out that the TSX way to do this is `let myDiv!: HTMLDivElement;` (we need the `!` because Solid generates an assignment, but ts doesn't know that)

<div ref={myDiv}>My Element</div>

```

This lets you access its properties and methods directly, allowing for manipulatinos such as animations and event handling.

> 

Reviewer: Would love a full component with an onMount(() => {console.log(ref)}) here

Author: could you please write this example out and we can put it in?

Reviewer: ```ts
const MyComponent = () => {
  let myDiv!: HTMLDivElement;

  onMount(() => {console.log(ref)});

  return <div ref={myDiv}>My Element</div>
}
```

## Callbacks

Refs can also take the form of a callback function.
This method can be convenient for encapsulating logic, especially when there is no need to wait until the elements are attached:

```tsx
let myDiv;

<div ref={(el) => myDiv = el}>My Element</div>
```

Callback refs provide a more flexible way to set refs and can be beneficial in various scenarious, such as when you need to attach event listeners to the DOM elements immediately after creation.

## Forwarding refs

Forwarding refs provide a way to expose one of the inner elements' references to a parent component.
By doing this, the parent component can directly access and interact the inner elements.

> When a component receives a `ref` attribute from its parent, it will come as `props.ref` in a callback form, regardless of whether the parent passed it as a simple assignment or a callback.

Reviewer: To help explain this, maybe also show ref={(el) => props.ref(el)} (which can be shortened to the example below)

Author: could you explicitly show me (I'm real out of it today tbf lol)

Reviewer: ```ts
const MyComponent = (props: {
  ref: HTMLDivElement | ((el: HTMLDivElement) => void);
}) => {
  return (
    <div ref={(el) => (props.ref as (el: HTMLDivElement) => void)(el)}>
      My Element
    </div>
  );
};

let myRef!: HTMLDivElement;

<MyComponent ref={myRef} />;
```

The component can then "forward" this ref to a desired internal element by assigning the `props.ref` directly to the `ref` attribute of that element.

Imagine there is a `Canvas` component, and from its parent, you want to directly manipulate the `canvas` element inside this Canvas component.

1. Canvas component

```tsx
function Canvas(props) {
  return (
    <div className="canvas-container">
      <canvas ref={props.ref} />
    </div>
  );
}
```

2. Parent component
```tsx
import { Canvas } from './Canvas.jsx';

function ParentComponent() {
  let canvasRef;

  const animateCanvas = () => {
    // Manipulate the canvas using canvasRef...
  };

  return (
    <div>
      <Canvas ref={canvasRef} />
      <button onClick={animateCanvas}>Animate Canvas</button>
    </div>
  );
}
```

In the `Canvas` component, we're directly assigning the incoming `props.ref` to the `ref` attribute of the internal canvas element. 
This forwards the reference to the parent component (`ParentComponent`), allowing it to have direct access to the `canvas` element.

## Directives

Directives allow the attachment of reusable behaviours to DOM elements.

> The [`use:`](/reference/use) prefix is used to denote these custom directives.

Reviewer: Consider adding something like this to this section:

Directives are like callback refs but they enable two extra features:
1. having multiple directives on an element
2. passing in reactive data to the callback

Unlike props or attributes, directives operate at a lower level through providing fine-grained control over the elements they are attached to.

A directive is essential a function with a specific signature:

```typescript
function directive(element: Element, accessor: () => any): void;
```

* `element`: This is the DOM element that the directive is being applied to.
* `accessor`: A function that gives access to the value(s) passed to the directive.

The directive functions are invoked at render time but are called before the element is added to the DOM.
Due to this order, elements are fully primed with their attributes, properties, or event listeners therefore minimizing unexpected behaviours or premature interactions.

Within directives, you're able to perform a variety of tasks, including:
* creating [signals](/essentials/signals)
* initiating [effects](/essentials/managing-side-effects)
* registering [cleanup tasks](/essentials/managing-side-effects#clean-up)
* adding [event listeners](#event-listeners)

> * and more.

Reviewer: Directives are pretty tricky to use with typescript without a fair bit of documentation, we should probably include something like this: https://www.solidjs.com/guides/typescript#use___

2. Using the Solid `ref` attribute

```jsx
const myDiv = <div>My Element</div>;
```

<Aside>
For a more idiomatic way, Solid provides the `ref` attribute.
Refs in Solid are assignments that occur at creation time before elements are attached to the document DOM.
This lets you access its properties and methods directly, allowing for manipulatinos such as animations and event handling.
To use `ref`, you simply need to declare a variable and use it as a ref attribute:

```tsx
let myDiv;

<div ref={myDiv}>My Element</div>

```

In TypeScript, you must use a definitive assignment assertion. 
This signals to TypeScript that `myDiv` will definitely be assigned, even if TypeScript can not confirm it.
This is particularly useful in Solid since the framework takes care of assigning the variable when the component is rendered.

```tsx
  let myDiv!: HTMLDivElement;
```
</Aside>

## Callbacks

Refs can also take the form of a callback function.
This method can be convenient for encapsulating logic, especially when there is no need to wait until the elements are attached:

```tsx
let myDiv;

<div ref={(el) => myDiv = el}>My Element</div>
```

> Callback refs provide a more flexible way to set refs and can be beneficial in various scenarious, such as when you need to attach event listeners to the DOM elements immediately after creation.

Reviewer 2: ```suggestion
Callback refs provide a more flexible way to set refs and can be beneficial in various scenarios, such as when you need to attach event listeners to the DOM elements immediately after creation.
```

import { Canvas } from './Canvas.jsx';

function ParentComponent() {
  let canvasRef;

  const animateCanvas = () => {
    // Manipulate the canvas using canvasRef...
  };

  return (
    <div>
      <Canvas ref={canvasRef} />
      <button onClick={animateCanvas}>Animate Canvas</button>
    </div>
  );
}
```

In the `Canvas` component, we're directly assigning the incoming `props.ref` to the `ref` attribute of the internal canvas element. 
This forwards the reference to the parent component (`ParentComponent`), allowing it to have direct access to the `canvas` element.

## Directives

Directives allow the attachment of reusable behaviours to DOM elements.
The [`use:`](/reference/use) prefix is used to denote these custom directives.
Unlike props or attributes, directives operate at a lower level through providing fine-grained control over the elements they are attached to.

Directives are like callback refs but they enable two extra features:
* Having multiple directives on an element.
* Passing in reactive data to the callback.

A directive is essentially a function with a specific signature:

```typescript
function directive(element: Element, accessor: () => any): void;
```

> * `element`: This is the DOM element that the directive is being applied to.

Reviewer 2: > is being applied to

Is there a better way to put this? 

Perhaps:

> The DOM element that the directive is applied to.

* `accessor`: A function that gives access to the value(s) passed to the directive.

> The directive functions are invoked at render time but are called before the element is added to the DOM.

Reviewer 2: ```suggestion
The directive functions are invoked at render time, but are called before the element is added to the DOM.
```

> Due to this order, elements are fully primed with their attributes, properties, or event listeners therefore minimizing unexpected behaviours or premature interactions.

Reviewer 2: ```suggestion
Due to this order, elements are fully primed with their attributes, properties, or event listeners, therefore minimizing unexpected behaviors or premature interactions.
```
